lib/AST/Printer/Printer.cpp

```diff
@@ -640,4 +718,469 @@ PrintedToken PrintedTokenRange::operator[](size_t index) const {
   return PrintedToken(impl, &(first[index]));
 }
 
+//namespace {
+//
+//static void Remap(
+//    std::vector<TokenContextImpl> &new_contexts,
+//    const std::vector<TokenContextImpl> &old_contexts,
+//    std::unordered_map<const void *, TokenContextIndex> &new_data_to_index,
+//    std::unordered_map<TokenContextIndex, TokenContextIndex> &index_remap) {
+//
+//}
+//
+//}  // namespace
+
+// Create a new printed token range by concatenating two printed token ranges
+// together.
+std::optional<PrintedTokenRange> PrintedTokenRange::Concatenate(
+    const PrintedTokenRange &a, const PrintedTokenRange &b) {
+  if (!a && !b) {
+    return std::nullopt;
+  } else if (!a) {
+    return b;
+  } else if (!b) {
+    return a;
+  }
+
+  if (&(a.impl->ast_context) != &(b.impl->ast_context)) {
+    return std::nullopt;
+  }
+
+  if (a.impl->ast.get() != b.impl->ast.get()) {
+    return std::nullopt;
+  }
+
+  auto new_impl = std::make_shared<PrintedTokenRangeImpl>(a.impl->ast_context);
+  new_impl->ast = a.impl->ast;
+  new_impl->tokens.reserve(a.impl->tokens.size() + b.impl->tokens.size() + 1u);
+  new_impl->contexts.reserve(
+      ((a.impl->contexts.size() + b.impl->contexts.size()) * 3u) / 2u);
+  new_impl->data = a.impl->data + b.impl->data;
```

Comment:
Why would I make a constructor to do this single-use thing? That just shifts where this code is; it doesn't magically make it better.

Reply:
Well yes the shifting is the point. Because as I am reading `Concatenate` why am I exposed impl details instead of some nicely named helper that hides it?
For example:
```
auto new_impl = PrintedTokenRangeImpl::MakeAsConcat( a, b );
```